src/json/json_parser.c

```diff
@@ -0,0 +1,260 @@
+                return parser_json_too_many_tokens;
+            case JSMN_ERROR_INVAL:
+                return parser_unexpected_characters;
+            case JSMN_ERROR_PART:
+                return parser_json_incomplete_json;
+            default:
+                return parser_json_unexpected_error;
+        }
+    }
+
+    parsed_json->numberOfTokens = 0;
+    parsed_json->isValid = 0;
+
+    // Parsing error
+    if (num_tokens <= 0) {
+        return parser_json_zero_tokens;
+    }
+
+    // We cannot support if number of tokens exceeds the limit
+    if (num_tokens > MAX_NUMBER_OF_TOKENS) {
+        return parser_json_too_many_tokens;
+    }
+
+    parsed_json->numberOfTokens = num_tokens;
+    parsed_json->isValid = true;
+
+    return parser_ok;
+}
+
+parser_error_t array_get_element_count(const parsed_json_t *json,
+                                       uint16_t array_token_index,
+                                       uint16_t *number_elements) {
+    *number_elements = 0;
+    if (array_token_index < 0 || array_token_index > json->numberOfTokens) {
+        return parser_no_data;
+    }
+
+    jsmntok_t array_token = json->tokens[array_token_index];
+    uint16_t token_index = array_token_index;
+    uint16_t prev_element_end = array_token.start;
+    while (true) {
+        token_index++;
+        if (token_index >= json->numberOfTokens) {
+            break;
+        }
+        jsmntok_t current_token = json->tokens[token_index];
+        if (current_token.start > array_token.end) {
+            break;
+        }
+        if (current_token.start <= prev_element_end) {
+            continue;
+        }
+        prev_element_end = current_token.end;
+        (*number_elements)++;
+    }
+
+    return parser_ok;
+}
+
+parser_error_t array_get_nth_element(const parsed_json_t *json,
+                                     uint16_t array_token_index,
+                                     uint16_t element_index,
+                                     uint16_t *token_index) {
+    if (array_token_index < 0 || array_token_index > json->numberOfTokens) {
+        return parser_no_data;
+    }
+
+    jsmntok_t array_token = json->tokens[array_token_index];
+    *token_index = array_token_index;
+
+    uint16_t element_count = 0;
+    uint16_t prev_element_end = array_token.start;
+    while (true) {
+        (*token_index)++;
+        if (*token_index >= json->numberOfTokens) {
+            break;
+        }
+        jsmntok_t current_token = json->tokens[*token_index];
+        if (current_token.start > array_token.end) {
+            break;
+        }
+        if (current_token.start <= prev_element_end) {
+            continue;
+        }
+        prev_element_end = current_token.end;
+        if (element_count == element_index) {
+            return parser_ok;
+        }
+        element_count++;
+    }
+
+    return parser_no_data;
+}
+
+parser_error_t object_get_element_count(const parsed_json_t *json,
```

Comment:
Lots of similarities with above function `array_get_nth_element`. Not sure if this can be optimized?

Reply:
Similar code but different purpose, not sure it is worth spending time refactoring it...